cirq-ft/cirq_ft/algos/arithmetic_gates.py

```diff
@@ -0,0 +1,386 @@
+    >>> N = 10
+    >>> data = [[1000 * x + 10 * y for y in range(x)] for x in range(N)]
+    >>> for i in range((N * (N - 1)) // 2):
+    ...     x = int(np.floor((1 + np.sqrt(1 + 8 * i)) / 2))
+    ...     y = i - (x * (x - 1)) // 2
+    ...     assert data[x][y] == 1000 * x + 10 * y
+
+     Note that both the for-loops iterate over the same ranges and in the same order. The only
+     difference is that the second loop is a "flattened" version of the first one.
+
+     Such a flattening of selection registers is useful when we want to load multi dimensional
+     data to a target register which is indexed on selection registers $x$ and $y$ such that
+     $0<= y <= x < N$ and we want to use a `SelectSwapQROM` to laod this data; which gives a
+     sqrt-speedup over a traditional QROM at the cost of using more memory and loading chunks
+     of size `sqrt(N)` in a single iteration. See the reference for more details.
+
+     References:
+         [Even More Efficient Quantum Computations of Chemistry Through Tensor Hypercontraction]
+         (https://arxiv.org/abs/2011.03494)
+            Lee et. al. (2020). Appendix F, Page 67.
+    """
+
+    bitsize: int
+    target_bitsize: int
+
+    def registers(self) -> Sequence[Union[int, Sequence[int]]]:
+        return [2] * self.bitsize, [2] * self.bitsize, [2] * self.target_bitsize
+
+    def with_registers(self, *new_registers) -> 'ContiguousRegisterGate':
+        x_bitsize, y_bitsize, target_bitsize = [len(reg) for reg in new_registers]
+        assert (
+            x_bitsize == y_bitsize
+        ), f'x_bitsize={x_bitsize} should be same as y_bitsize={y_bitsize}'
+        return ContiguousRegisterGate(x_bitsize, target_bitsize)
+
+    def apply(self, *register_vals: int) -> Union[int, Iterable[int]]:
+        x, y, target = register_vals
+        return x, y, target ^ ((x * (x - 1)) // 2 + y)
+
+    def _circuit_diagram_info_(self, _) -> cirq.CircuitDiagramInfo:
+        wire_symbols = ["In(x)"] * self.bitsize
+        wire_symbols += ["In(y)"] * self.bitsize
+        wire_symbols += ['+(x(x-1)/2 + y)'] * self.target_bitsize
+        return cirq.CircuitDiagramInfo(wire_symbols=wire_symbols)
+
+    def _t_complexity_(self) -> infra.TComplexity:
+        # See the linked reference for explanation of the Toffoli complexity.
+        toffoli_complexity = infra.t_complexity(cirq.CCNOT)
+        return (self.bitsize**2 + self.bitsize - 1) * toffoli_complexity
+
+    def __repr__(self) -> str:
+        return f'cirq_ft.ContiguousRegisterGate({self.bitsize}, {self.target_bitsize})'
+
+    def __pow__(self, power: int):
+        if power in [1, -1]:
+            return self
+        return NotImplemented  # coverage: ignore
+
+
+@attr.frozen
+class AdditionGate(cirq.ArithmeticGate):
+    """Applies U|p>|q> -> |p>|p+q>.
+
+    Args:
+        bitsize: The number of bits used to represent each integer p and q.
+            Note that this adder does not detect overflow if bitsize is not
```

Comment:
What does happen if there is overflow?  Is the largest bit dropped?

Reply:
Yes, expanded the docstring.